rpcclient/roundrobin.go

```diff
@@ -0,0 +1,347 @@
+type Balancer interface {
+	//Get gets the next usabel connection
+	Get() (wsConn *websocket.Conn, hostAddress *HostAddress, err error)
+	//GetConnectionState gets the connectionstate of corresponding wsConn
+	GetConnectionState(host string) (state ConnectionState, ok bool)
+	// GetWsConnection gets the wsConn corresponding to the host
+	GetWsConnection(host string) (wsConn *websocket.Conn, connOk bool)
+	// NotifyConnStateChange is called by rpcClient when the ConnectionState
+	// for a hostaddress changes.
+	NotifyConnStateChange(sc *HostAddress, state ConnectionState, sync bool)
+	//UpdateReconnectAttempt will increase retryattempt + 1
+	UpdateReconnectAttempt(hostAdd *HostAddress)
+	//NotifyReconnect will update the map for wsConns and update the connection state
+	//for this address
+	NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress)
+	// Close closes corresponding connection if host is not empty
+	// else would close all the WS connections, setting the state as Disconnected
+	Close(host string)
+}
+
+// Picker is used to pick the next connection to be used
+type Picker interface {
+	// Pick checks for the next usable connection using the connState map with balancer
+	Pick(balancer Balancer) (*HostAddress, error)
+}
+
+type picker struct {
+	conns []HostAddress
+	mu    sync.Mutex
+	next  int
+}
+
+//Pick checks for the next usable connection using the connState map
+//It changes the state to Conecting, if the connection picked is not yet used
+func (p *picker) Pick(balancer Balancer) (conn *HostAddress, err error) {
+
+	if len(p.conns) <= 0 {
+		log.Infof(ErrNoConnAvailable.Error())
+		return conn, ErrNoConnAvailable
+	}
+
+	p.mu.Lock()
+	startPos := p.next
+	var ok bool
+
+	for {
+		conn = &p.conns[p.next]
+		state, ok := balancer.GetConnectionState(conn.Host)
+		p.next = (p.next + 1) % len(p.conns)
+		if !ok || state == Ready || state == Idle {
+			//either not yet tried or ready to use
+			break
+		}
+		//back to start
+		if startPos == p.next {
+			break
+		}
+	}
+	if conn == nil {
+		return conn, ErrNoUsableConnAvailable
+	}
+	if !ok {
+		balancer.NotifyConnStateChange(conn, Connecting, true)
+	}
+	p.mu.Unlock()
+	return conn, nil
+}
+
+// RoundRobinBalancer represents the balancer implementing Balancer
+// and maintaining a Picker
+type RoundRobinBalancer struct {
+	connPicker    picker
+	connConfig    *ConnConfig
+	wsConns       map[string]*websocket.Conn
+	hostAddMap    map[string]*HostAddress
+	mu            sync.Mutex
+	connState     map[string]ConnectionState
+	waitToConnect chan struct{}
+	isReady       bool
+	// This is used to decide if client.start() needs to be called during reconnect
+	// Set to true when all connection are either disconnected or shutdown
+	// Set to false after client.start()
+	NeedsClientRestart bool
+}
+
+//Get gets the next connection to be used.
+//It returns both *websocket.Conn and *HostAddress
+func (rrb *RoundRobinBalancer) Get() (*websocket.Conn, *HostAddress, error) {
```

Comment:
I would call this `NextWsConn` since it is more descriptive and doesn't run afoul of the effective Go guidelines mentioned further down in the review.

Reply:
Renamed it to NextConn since wsConn may be nil in case of HttpPostMode.